net.solarnetwork.node.setup.web/web/js/graph-data.js

```diff
@@ -0,0 +1,237 @@
+//Author Robert Rewcastle
+//script to draw line graphs of datums with a wattage variable
+//this script will put the graphs inside an html tag with a graphpoint class which 
+//one must put in the html document before loading the script
+//requires d3 https://d3js.org/d3.v3.js
+
+//json field we are interested in
+var key = "watts";
+var units = "W";
+
+//This map is graphs to look up the latest reading based on their sourceId
+var datamap = {};
+
+//this array will contain all the sourceIds of graphs we are plotting
+var sourceIds = [];
+
+//https://stackoverflow.com/questions/1960473/get-all-unique-values-in-an-array-remove-duplicates
+//This get used for deciding how many ticks on the y axis
+function onlyUnique(value, index, self) { 
+    return self.indexOf(value) === index;
+}
+
+//https://stackoverflow.com/questions/7343890/standard-deviation-javascript
+//This get used for giving the graph some white space either side of the graph
+function StandardDeviation(numbersArr) {
+    //--CALCULATE AVAREGE--
+    var total = 0;
+    for(var key in numbersArr) 
+       total += numbersArr[key];
+    var meanVal = total / numbersArr.length;
+    //--CALCULATE AVAREGE--
+  
+    //--CALCULATE STANDARD DEVIATION--
+    var SDprep = 0;
+    for(var key in numbersArr) 
+       SDprep += Math.pow((parseFloat(numbersArr[key]) - meanVal),2);
+    var SDresult = Math.sqrt(SDprep/numbersArr.length);
+    //--CALCULATE STANDARD DEVIATION--
+    return SDresult
+    
+}
+
+//when a new datum comes in, this handler gets called
+//the handler checks if the new datum has a wattage reading and 
+var handler = function handleMessage(msg) {
+	
+	var datum = JSON.parse(msg.body).data;
+
+	//check that the datum has a reading
+	if (datum[key] != undefined){
+		//map the sourceId to the current reading
+		datamap[datum.sourceId] = datum[key];
+		
+		//if we have not seen this sourceId before we need to graph it
+		if (sourceIds.indexOf(datum.sourceId)==-1){
```

Comment:
I think the entire `sourceIds` variable isn't necessary, because `if ( datamap[datum.sourceId] )` would provide the same answer as `if (sourceIds.indexOf(datum.sourceId)==-1)`.